bliss/catalog.py

```diff
@@ -722,3 +722,35 @@ def to_astropy_table(self, encoder_survey_bands: Tuple[str]) -> Table:
         est_cat_table.remove_column("galaxy_params")
 
         return hstack([est_cat_table, galaxy_params_table])
+
+    def filter_full_catalog_by_ploc_box(self, box_origin: torch.Tensor, box_len: float):
+        assert box_origin[0] + box_len < self.height, "invalid box"
+        assert box_origin[1] + box_len < self.width, "invalid box"
+
+        box_origin_tensor = box_origin.view(1, 1, 2).to(device=self.device)
+        box_end_tensor = (box_origin + box_len).view(1, 1, 2).to(device=self.device)
+
+        plocs_mask = torch.all(
+            (self.plocs < box_end_tensor) & (self.plocs > box_origin_tensor), dim=2
+        )
+
+        plocs_mask_indexes = plocs_mask.nonzero()
+        plocs_inverse_mask_indexes = (~plocs_mask).nonzero()
+        plocs_full_mask_indexes = torch.cat((plocs_mask_indexes, plocs_inverse_mask_indexes), dim=0)
+        _, index_order = plocs_full_mask_indexes[:, 0].sort(stable=True)
+        plocs_full_mask_sorted_indexes = plocs_full_mask_indexes[index_order.tolist(), :]
+
+        d = {}
+        new_max_sources = plocs_mask.sum(dim=1).max()
+        for k, v in self.to_dict().items():
+            if k == "n_sources":
+                d[k] = plocs_mask.sum(dim=1)
+            else:
+                d[k] = v[
+                    plocs_full_mask_sorted_indexes[:, 0].tolist(),
+                    plocs_full_mask_sorted_indexes[:, 1].tolist(),
+                ].view(-1, self.max_sources, v.shape[-1])[:, :new_max_sources, :]
```

Comment:
This code seems cryptic to me. Can you use some intermediate variables to better explain what it's doing? (Is a tensor getting converted to a Python list and then back to a tensor?)

Reply:
I'm trying to do the following operation.
```
>>> t.shape
torch.Size([4, 5, 2])
>>> t[[0, 1], [3, 4]]
tensor([[0.9178, 0.3564],
        [0.4694, 0.9806]])
>>> t[[1, 2], [3, 4]]
tensor([[0.0147, 0.4497],
        [0.2121, 0.2853]])
>>> 
```

Reply:
the tensor `plocs_full_mask_sorted_indexes` contain such indexes, but i find it will throw error if i use `v[plocs_full_mask_sorted_indexes]`.

Reply:
OK I see